<h1 style="text-align: center;">Linear Regression Analysis</h1>
<h1 style="text-align: center;">Homework 3</h1>
<h3 style="text-align: center;">Brandon Miner</h3>
<h3 style="text-align: center;">September 10, 2025</h3>

### Written Questions

1. Set up the $X$ matrix and $\beta$ vector for each of the following regression models (assume $i=1,\dots,5$):

a. $Y_i=\beta_1X_{i1}+\beta_2X_{i2}+\beta_3X_{i1}^2+\epsilon_i$

$$
X = \begin{bmatrix}
1 & x_{11} & x_{12} & x_{11}^2 \\
1 & x_{21} & x_{22} & x_{21}^2 \\
1 & x_{31} & x_{32} & x_{31}^2 \\
1 & x_{41} & x_{42} & x_{41}^2 \\
1 & x_{51} & x_{52} & x_{51}^2 \\
\end{bmatrix}, \beta = \begin{bmatrix}
0 \\ \beta_1 \\ \beta_2 \\ \beta_3
\end{bmatrix},
\hat Y = X\beta
$$

b. $\sqrt{Y_i}=\beta_0+\beta_1X_{i1}+\beta_2\log_{10}X_{i2}+\epsilon_i$

$$
X = \begin{bmatrix}
1 & x_{11} & \log{x_{12}} \\
1 & x_{21} & \log{x_{22}} \\
1 & x_{31} & \log{x_{32}} \\
1 & x_{41} & \log{x_{42}} \\
1 & x_{51} & \log{x_{52}} \\
\end{bmatrix}, 
\beta = \begin{bmatrix} \beta_0 \\ \beta_1 \\ \beta_2 \end{bmatrix}, 
\hat Y = \beta^TX^TX\beta
$$

2. A student stated: “Adding predictor variables to a regression model can never reduce $R^2$, so we should include all available predictor variables in the model." Comment

This actually highlights the issue with $R^2$ which is artificially increases as you add predictors, but this doesn't actually mean the model is more accurate or more precise.
Adjusted $R^2$ exists for the sole purpose of fixing this issue, to make $R^2$ practically useful for diagnostics.

3. Consider the multiple regression model: 
$$Y_i=\beta_1X_{i1}+\beta_2x_{i2}+\epsilon_i,\qquad i=1,\dots,n$$
where the $\epsilon_i$ are uncorrelated, with $E(\epsilon_1)=0$ and $\text{Var}(\epsilon_1)=\sigma^2$

a. State the least squares criterion and derive the least squares estimators of $\beta_1$ and $\beta_2$.

The least squares criterion is the $\stackrel{\text{argmin}}{_{_{_{_\beta}}}}\mathbb Q(\beta)$ where $\mathbb Q(\beta) = ||Y-X\beta||^2$.
$$\frac{\partial \mathbb Q}{\partial\beta}=\frac{\partial}{\partial\beta}(||Y-X\beta||^2)=\frac{\partial}{\partial\beta}((Y-X\beta)^T(Y-X\beta))$$
$$=\frac{\partial}{\partial\beta}(Y^TY-Y^TX\beta-(X\beta)^TY+(X\beta)^T(X\beta))$$
$$=\frac{\partial}{\partial\beta}(Y^TY-2\beta^TX^TY+\beta^TX^TX\beta)$$
$$=-2(X^TY)+2(X^TX)\beta\stackrel{!}{=}0$$
Solving for $\beta$
$$\hat\beta=(X^TX)^{-1}X^TY$$

B. Assuming that $\epsilon_i$ are independent normal random variables, state the likelihood function and obtain the maximum likelihood estimators of $\beta_1$ and $\beta_2$. Are these the same as the least squares estimators?

$$ L(\beta | X) = \prod^n_{i=1}\mathbb Q(\beta_i)$$
$$\ell(\beta | X) = \sum^n_{i=1}\mathbb Q(\beta_i)$$
$$ = \sum^n_{i=1} ||Y_i-X\beta_i||^2$$
$$ = \sum^n_{i=1} (Y_i-X\beta_i)T(Y-X\beta_i)$$
$$ = \sum^n_{i=1} Y_i^TY_i - 2\beta_iX^TY_i + \beta_i^TX^TX\beta_i$$
$$ \frac{\partial\ell}{\partial\beta} = \sum^n_{i=1} -2(X^TY_i)+2(X^TX)\beta_i\stackrel!=0$$
$$2\sum^n_{i=1}(X^TX)\beta_i = 2\sum^n_{i=1}X^TY_i$$
$$\sum^n_{i=1}\beta_i = (X^TX)^{-1}X^T\sum^n_{i=1} Y_i$$
$$\hat\beta = (X^TX)^{-1}X^TY$$

From here we can easily see that the estimators are equivalent.

4. An analyst wanted to fit the regression model
$$Y_i=\beta_0+\beta_1X_{i1}+\beta_2X_{i2}+\beta_3x_{i3}+\epsilon_i,\qquad i=1,\dots,n$$
by the method of least squares when it is known that $\beta_2=4$. How can the analyst obtain the desired fit by using a multiple regression computer program?

Instead of regressing $X_1,X_2,X_3$ onto Y, we can define a new column $Y-\beta_2X_2$ and regress $X_1,X_3$ onto $Y-\beta_2X_2$. This works because our new model we are trying to estimate is $$Y_i-\beta_2X_{i2} = \beta_0 + \beta_1X_{i1} + \beta_3X_{i3}+\epsilon_i$$
Since we know $\beta_2=4$, we can actually derive $Y_i-4X_{i2}$.

Code concept:
```python
data  # dataframe of Y,X_1,X_2,X_3
data["Y-beta_2"] = data["Y"] - 4*data["X_2"]

model = smf.ols("Y-beta_2~X_1+X_3").fit()
```

5. In a small-scale regression study, the following data were obtained:

| $i$ | 1 | 2 | 3 | 4 | 5 | 6 |
| - | - | - | - | - | - | - |
| $X_{i1}$ | 7 | 4 | 16 | 3 | 21 | 8 |
| $X_{i2}$ | 33 | 41 | 7 | 49 | 5 | 31 |
| $Y_i$ | 42 | 33 | 75 | 28 | 91 | 55 |

Assume that regression model with independent normal error terms is appropriate. Using matrix
methods, obtain: (a) $\hat\beta$; (b) $e$; (c) $H$; (d) $SSR$; (e) $\widehat{SE}(\hat\beta)$; (f) $\hat Y_h$ when $X_{h1}=10$, $X_{h2}=30$; and (g) $\widehat{SE}(\hat Y_h)$ when $X_{h1}=10$, $X_{h2}=30$.

Write down matrix $X$ and vector $y$, and the formula to calculate each value before giving the answers. You can do the matrix calculation using Python matrix multiplication or any calculator. Do NOT use Python packages to fit the model and read the output directly.

In [1]:
import numpy as np

(a) 
$$\hat\beta = (X^TX)^{-1}X^TY$$

In [17]:
X = np.array( [[ 7, 4,16, 3,21, 8],
               [33,41, 7,49, 5,31]])
X = X.T
Y = np.array(  [42,33,75,28,91,55])
beta_hat = np.linalg.inv(X.T @ X) @ X.T @ Y
print(f"beta_hat is {beta_hat}")

beta_hat is [4.40479026 0.38033777]


(b)
$$
e = Y - \hat Y = Y - X\hat\beta = Y-X(X^TX)^{-1}X^TY
$$

In [18]:
Y_hat = X @ beta_hat
e = Y - Y_hat
print(f"the residuals are {e}")

the residuals are [-1.38467826 -0.21300967  1.8609915  -3.85092159 -3.40228423  7.97120702]


(c)
$$
H = X\hat\beta=X(X^TX)^{-1}X^TY
$$

In [19]:
H = X @ beta_hat
print(f"The hat matrix is {H}")

The hat matrix is [43.38467826 33.21300967 73.1390085  31.85092159 94.40228423 47.02879298]


(d)
$$
SSR = Y - \bar y = Y - \frac1n\sum^n_{i=1}y_i
$$

In [21]:
SSR = sum((Y-np.mean(Y))**2)
print(f"The SSR is {SSR}")

The SSR is 3072.0


(e)
$$
\widehat{SE} (\hat\beta) = \sqrt{\widehat{\text{Var}}(\hat\beta)} = \sqrt{\hat\sigma^2(X^TX)^{-1}} = \hat\sigma(X^TX)^{-1/2}
$$
where $$\hat\sigma^2=\frac{SSE}{n-p}$$

In [25]:
SSE = e.T @ e
sigma_squared = SSE/(len(Y)-len(X.T))
std_err_beta_hat = sigma_squared * (X.T @ X)**(-.5)
print(f"The standard Error of beta_hat is\n{std_err_beta_hat[0]}\n{std_err_beta_hat[1]}")

The standard Error of beta_hat is
[0.82511475 0.75135097]
[0.75135097 0.30265768]


(f)
$$
\hat Y_h = X_h\hat\beta = \begin{bmatrix} 10 & 30 \end{bmatrix}\hat\beta
$$

In [26]:
X_h = np.array([10,30])
Y_hat_h = X_h @ beta_hat
print(f"Y_hat_h is {Y_hat_h}")

Y_hat_h is 55.45803571578524


(g)
$$
\widehat{SE}(\hat Y_h) = \widehat{SE}(\beta_{h1}X_{h1} + \beta_{h2}X_{h2}+\epsilon_h) = \sqrt{\widehat{\text{Var}}(\epsilon_h)} = \hat\sigma
$$

In [27]:
print(f"Variance, and therefore standard error is fixed for the fitted values to be the same as that of the error term, hence {sigma_squared}")

Variance, and therefore standard error is fixed for the fitted values to be the same as that of the error term, hence 23.842818226945976


6. Consider the multiple linear regression model in the matrix form $Y = X\beta + \epsilon$ with 
$E(\epsilon) = 0$ and $\text{Var}(\epsilon) = \sigma^2I_n$. Let $\hat\beta$ be the least squares estimator of $\beta$, $\hat Y = X \hat\beta$ be the fitted values, and $e = Y − \hat Y$ be the residuals. Express $\text{Var}( \hat\beta)$, $\text{Var}( \hat Y )$, and $\text{Var}(e)$ in terms of $\sigma^2$ and $X$. Show your steps to obtain the expressions in detail.

$\text{Var}(\hat\beta)$: Given that
$$Y\sim N(X\beta,\sigma^2I)$$

$$(X^TX)^{-1}X^TY\sim N((X^TX)^{-1}X^TX\beta, \sigma^2(X^TX)^{-1}X^TI)$$

$$\hat\beta\sim N (\beta, \sigma^2(X^TX)^{-1}X^T)$$

Therefore $\text{Var}(\hat\beta) = \sigma^2(X^TX)^{-1}X^T$

$\text{Var}(\hat Y)$: Given that

$$\hat\beta\sim N (\beta, \sigma^2(X^TX)^{-1}X^T)$$

$$X\hat\beta\sim N(X\beta, \sigma^2H)$$

$$\hat Y\sim N(X\beta, \sigma^2H)$$

Therefore $\text{Var}(\hat Y) = \sigma^2H$

$\text{Var}(e)$: Given that
$$Y\sim N(X\beta,\sigma^2I),\space \hat Y\sim N(X\beta, \sigma^2H)$$

$$Y-\hat Y \sim N ( X\beta - X\beta , \sigma^2I - \sigma^2H)$$

$$e \sim N (0, \sigma^2(I - H))$$

Therefore $\text{Var}(e) = \sigma^2(I - H)$





### Coding Questions

In [59]:
import statsmodels.formula.api as smf
import statsmodels.api as sm
from scipy import stats
import pandas as pd


1. The dataset "insurance.csv" contains information on the individuals’ medical costs billed by health
insurance with the variables defined as below

- age: age of primary beneficiary
- sex: insurance contractor gender, female, male
- bmi: Body mass index, providing an understanding of weights that are relatively high or low relative to height, objective index of body weight (kg/m2) using the ratio of height to weight
- children: Number of children covered by health insurance / Number of dependents
- smoker: Smoking
- region: the beneficiary’s residential area in the US, northeast, southeast, southwest, northwest.
- charges: Individual medical costs billed by health insurance

Regress "charges" against "age", "bmi" and "children" (use the order: charges ∼age + bmi + childern) in Python

In [12]:
data = pd.read_csv("../../data/insurance.csv")
data.head(3)

age     sex    bmi  children smoker     region     charges
0   19  female  27.90         0    yes  southwest  16884.9240
1   18    male  33.77         1     no  southeast   1725.5523
2   28    male  33.00         3     no  southeast   4449.4620

(a) ONLY based on the model summary table (`model.summary()`), we have discussed several important things to look at when it comes to the model interpretation: Global F-test p-value, $R^2$
and $R^2_a$, and t-test for individual coefficient. Comment on each values above from your model summary, what kind of conclusion you have about your model fit and significance of predictors so far

In [4]:
model = smf.ols("charges~age+bmi+children", data=data).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                charges   R-squared:                       0.120
Model:                            OLS   Adj. R-squared:                  0.118
Method:                 Least Squares   F-statistic:                     60.69
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           8.80e-37
Time:                        18:40:33   Log-Likelihood:                -14392.
No. Observations:                1338   AIC:                         2.879e+04
Df Residuals:                    1334   BIC:                         2.881e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -6916.2433   1757.480     -3.935      0.000   -1.04e+04   -3468.518
age          239.9945     22.289     10.767      0.000     196.269     283.720
bmi          332.0834     51.310      6.472      0.000     231.425     432.741
children     542.8647    258.241      2.102      0.036      36.261    1049.468
==============================================================================
Omnibus:                      325.395   Durbin-Watson:                   2.012
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              603.372
Skew:                           1.520   Prob(JB):                    9.54e-132
Kurtosis:                       4.255   Cond. No.                         290.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Both the $R^2$ and $R^2_a$ are extremely low which implies there is a lot of unexplained variance in the data, which implies the model is not a great predictor of the model. The global F-test p-value is very low which means that there is clearly a correlation between our predictors and response variable. Putting both pieces of evidence together, the predictors seem to be strong predictors of the response, but not a linear model.

(b) Run `sm.stats.anova_lm(model,typ=1)` to obtain sum_sq of “bmi” in the ANOVA is a sequential
sum of squares of “bmi” which is defined as $SeqSS(bmi)=SSEreduced−SSEfull$.
What is the reduced model and full model here? How do we interpret the p-value of “bmi” in
this F-test?

In [9]:
sm.stats.anova_lm(model,typ=1)


df        sum_sq       mean_sq           F        PR(>F)
age          1.0  1.753019e+10  1.753019e+10  135.546341  6.627851e-30
bmi          1.0  5.446449e+09  5.446449e+09   42.112843  1.211545e-10
children     1.0  5.715190e+08  5.715190e+08    4.419080  3.572625e-02
Residual  1334.0  1.725261e+11  1.293299e+08         NaN           NaN

The reduced model for "bmi" in the Sequential ANOVA is: $charges = -6916.2433 + 239.9945age$

The full model for "bmi" in the Sequential ANOVA is: $charges = -6916.2433 + 239.9945age + 332.0834bmi$

The p-value is the probability that "bmi" isn't a significant predictor of "charges" when compared to a model with just "age" as a predictor.

(c) Run `sm.stats.anova_lm(model,typ=2)` to obtain sum_sq of “bmi” in the anova is a partial sum of squares of “bmi” defined by $PartialSS(bmi) = SSEreduced −SSEfull$.
What is the reduced model and full model here? How do we interpret the p-value of “bmi” in this F-test

In [10]:
sm.stats.anova_lm(model,typ=2)

sum_sq      df           F        PR(>F)
age       1.499426e+10     1.0  115.938067  5.533923e-26
bmi       5.417280e+09     1.0   41.887301  1.354882e-10
children  5.715190e+08     1.0    4.419080  3.572625e-02
Residual  1.725261e+11  1334.0         NaN           NaN

The reduced model for "bmi" in the Sequential ANOVA is: $charges = 239.9945age + 542.8647children$

The full model for "bmi" in the Sequential ANOVA is: $charges = 239.9945age + 332.0834bmi + + 542.8647children$

The p-value is the probability that "bmi" isn't a significant predictor of "charges" when compared to a model of all other predictors.

2. A commercial real estate company evaluates age ($X_1$), operating expenses ($X_2$, in thousands of dollars), vacancy rate ($X_3$), total square footage ($X_4$) and rental rates ($Y$ , in thousands of dollars) for commercial properties in a large metropolitan area in order to provide clients with quantitative information upon which to make rental decisions. The data are taken from 81 suburban commercial properties. (The data can be found in `property.txt`; The first column is $Y$ , followed by $X_1, X_2, X_3, X_4$.) Consider the multiple linear regression model

$$Y_i=\beta_0+\beta_1X_{1i}+\beta_2X_{2i}+\beta_3X_{3i}+\beta_4X_{4i}+\epsilon_i,i=1,\dots,n,\text{ with }\epsilon_i\stackrel{\text{iid}}{\sim} N(0,\sigma^2)$$

In [29]:
data = pd.read_csv('../../data/property.txt', 
                   delim_whitespace=True,
                   header=None, 
                   names=["rental_rates","age","expenses","vacancy_rate","square_footage"])
data.head(3)

/var/folders/09/4h25jfnn06q73576vknj8hz40000gn/T/ipykernel_23169/1786286897.py:1: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  data = pd.read_csv('../../data/property.txt',


rental_rates  age  expenses  vacancy_rate  square_footage
0          13.5    1      5.02          0.14          123000
1          12.0   14      8.19          0.27          104079
2          10.5   16      3.00          0.00           39998

(a) Fit a model in Python and produce the summary table.

In [31]:
model = smf.ols("rental_rates~age+expenses+vacancy_rate+square_footage",data=data).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           rental_rates   R-squared:                       0.585
Model:                            OLS   Adj. R-squared:                  0.563
Method:                 Least Squares   F-statistic:                     26.76
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           7.27e-14
Time:                        19:21:16   Log-Likelihood:                -122.75
No. Observations:                  81   AIC:                             255.5
Df Residuals:                      76   BIC:                             267.5
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         12.2006      0.578     21.110      0.000      11.049      13.352
age               -0.1420      0.021     -6.655      0.000      -0.185      -0.100
expenses           0.2820      0.063      4.464      0.000       0.156       0.408
vacancy_rate       0.6193      1.087      0.570      0.570      -1.545       2.784
square_footage  7.924e-06   1.38e-06      5.722      0.000    5.17e-06    1.07e-05
==============================================================================
Omnibus:                        1.922   Durbin-Watson:                   1.580
Prob(Omnibus):                  0.383   Jarque-Bera (JB):                1.301
Skew:                           0.148   Prob(JB):                        0.522
Kurtosis:                       3.545   Cond. No.                     1.74e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.74e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

(b) Compute the fitted values and residuals. Print the first 6 cases. Based on the residuals, give an estimate of  $\sigma^2$.

In [60]:
fitted_values = model.fittedvalues[:5]
residuals = model.resid[:5]

fitval_resid_df = pd.DataFrame({
    "Fitted Values" : fitted_values,
    "Residuals" : residuals
})

SSE = sum(residuals**2)

display(fitval_resid_df)

print(
    f"sigma-squared-hat = {\
        round(SSE / (model.df_resid), 
            5)}")


Fitted Values  Residuals
0      14.535672  -1.035672
1      13.513806  -1.513806
2      11.091053  -0.591053
3      15.133568  -0.133568
4      13.686716   0.313284

sigma-squared-hat = 0.05039


(c) Test whether $\beta_2 = 0$ or not at 0.01 significance level. Interpret this coefficient in the real life terms of the problem.

In [74]:
beta_2_t_value = model.tvalues['expenses']
crit_val = stats.t.cdf(1-0.01/2,model.df_resid)

if beta_2_t_value > crit_val:
    print(f"test-statistic = {abs(beta_2_t_value):.3f} > {abs(crit_val):.3f} = critical-value")
    print("We reject the null hypothesis. Expenses (beta_2) is a statistically significant predictor of rental rates at a 0.01 significance level.")
else:
    print(f"test-statistic = {abs(beta_2_t_value):.3f} < {abs(crit_val):.3f} = critical-value")
    print("We fail to reject the null hypothesis. We cannot say that expenses (beta_2) is a statistically significant predictor of rental rates at a 0.01 significance level.")


test-statistic = 4.464 > 0.839 = critical-value
We reject the null hypothesis. Expenses (beta_2) is a statistically significant predictor of rental rates at a 0.01 significance level.


Test $H_0:\beta_1=\beta_2=0$ vs $H_a:$ either $\beta_1$ or $\beta_2$ not equal to 0.

In [79]:
f_statistic = model.fvalue
f_pvalue = model.f_pvalue

print(f"F-statistic: {f_statistic:.4f}")
print(f"P-value of F-test: {f_pvalue}")

if f_pvalue < 0.05:
    print("We reject the null hypothesis. At least one parameter is a significant predictor of charges at a 0.5 level of significance.")
else:
    print("We fail to reject the null hypothesis. We cannot say that at least one parameter is a significant predictor of charges at a 0.5 level of significance.")


F-statistic: 26.7555
P-value of F-test: 7.271987454038193e-14
We reject the null hypothesis. At least one parameter is a significant predictor of charges at a 0.5 level of significance.


(e) Consider a property with the following characteristics: $X_1 = 4,X_2 = 10,X_3 = 0.1,X_4 = 80000$. Give a 90% prediction interval for the rental rate.

In [88]:
pred_obs = pd.DataFrame({
    "age" : [4],
    "expenses" : [10],
    "vacancy_rate" : [0.1],
    "square_footage" : [80000]
})

pred_stat = model.get_prediction(pred_obs)

prediction_interval = pred_stat.summary_frame(alpha=0.05)
pred_lower = prediction_interval['obs_ci_lower'][0]
pred_upper = prediction_interval['obs_ci_upper'][0]

print("Given a house that is:")
print("- 4 years old")
print("- 10 thousand dollars of operating expenses")
print("- 0.1 vacancy rate")
print("- 80,000 sq footage")
print(f"\nThe prediction interval is [{pred_lower:.3f},{pred_upper:.3f}]")

Given a house that is:
- 4 years old
- 10 thousand dollars of operating expenses
- 0.1 vacancy rate
- 80,000 sq footage

The prediction interval is [12.852,17.444]
